text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+# OpenTelemetry versioning and stability
+
+OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability gurantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.
+
+Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this corss-language proposal, and produce a language-specific proposal which details how these requirements will be met.
+
+## Design goals
+
+**Provide end users with coherent, tested releases.**  
+We do not want users to run untested combinations of core packages. Users must be able to say “I am using OpenTelemetry v1.2.0 for Python” and have a shared understanding as to what that means.
+
+**Ensure that end users stay up to date with the latest release.**  
+We want all users to  stay up to date with the latest version of OpenTelemetry. We do not want to create hard breaks in support, of any kind, which leave users stranded on older versions. It must always be possible to upgrade to the latest minor version of OpenTelemetry, without creating a compilation error.
+
+**Never create a dependency conflict between packages which rely on different versions of OpenTelemetry. Avoid breaking all stable public APIs.**  
+Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years.
+
+**Allow for multiple levels of package stability within the same release.**  
+Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.
+
+## Relevant architecure
+
+OpenTelemetry is split into signals. Each signal provides a specialized form of observability.
+
+Each signal consists of four types of packages:
+
+**API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with  OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.
+
+**SDK -** As the implemention the API. The SDK is only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.
```

Comment:
```suggestion
**SDK -** As the implementation the API, the SDK is generally only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.
```

Reply:
SDKs generally (I think?) provide dynamic tracing configuration (sampling configuration, for instance), so they will be referenced during the ongoing running of the application if that feature is used.

Reply:
To be clear, ongoing but in the background, correct? Not within app or library code as a dependecy?

Reply:
I'm not sure. If we provide the SDK APIs to update tracing configuration, this could be done by any application in any way that they want to. Not by instrumentation, surely, but definitely by application code. The zpages extensions do, this, for example.

Reply:
Well, if we are creating SDK interfaces and are encouraging out end users to take dependencies on them in application code, AKA within transactions or mixed in with business logic, or encouraging shared libraries to make calls directly to the SDK, then that is a big problem. That code is no longer "SDK" code, that is part of our public interface, and must be put into the API. 

Is that the issue here, or are you saying there may be background processes within an application which run periodically, but are only managed at setup and do not get mixed into application code as a cross cutting concern?

Reply:
I think what you're saying in the comment makes sense, but I don't think that means "only referenced during setup and teardown". What you're saying is that "application code involving transactions and business logic" shouldn't depend on the SDK, which I agree with. However, operator-oriented code will and should probably use the SDK to fulfill their more dynamic needs around how the SDK is functioning.

Reply:
Ok thanks, I'll clarify this.